File: TS/hellointerview/intervals/mergeIntervals.ts
class Solution {
  mergeIntervals(intervals: number[][]): number[][] {
    const merged: number[][] = [];
    // check 0 or 1 length edge case

    intervals.sort((a, b) => a[0] - b[0]);
    const n = intervals.length;

    for (const interval of intervals) {
      if (merged.length === 0 || interval[0] > merged[merged.length - 1][1]) {
        merged.push(interval);
      } else {
        merged[merged.length - 1][1] = Math.max(merged[merged.length - 1][1], interval[1]);
      }
    }

    return merged;
  }
}

File: Java/hellointerview/03-intervals/mergeIntervals.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
  public int[][] mergeIntervals(int[][] intervals) {
    Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
    List<int[]> merged = new ArrayList<>();

    for (int[] interval : intervals) {
      if (merged.isEmpty() || interval[0] > merged.get(merged.size() - 1)[1]) {
        // intervals don't overlap
        merged.add(interval);
      } else {
        // intervals overlap
        merged.get(merged.size() - 1)[1] = Math.max(interval[1], merged.get(merged.size() - 1)[1]);
      }
    }

    return merged.toArray(new int[merged.size()]);
  }
}
